In [6]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

#### Q1.  Who are the top 5 users in Boston with the highest number of followers? List their login in order, comma-separated

In [7]:
users = pd.read_csv('users.csv')
users.head()

login               name              company    location  \
0               brianyu28           Brian Yu                  NaN  Boston, MA   
1                  nzakas  Nicholas C. Zakas  HUMAN WHO CODES LLC  Boston, MA   
2           PatrickAlphaC    Patrick Collins               CYFRIN      Boston   
3              KeithGalli        Keith Galli                  NaN  Boston, MA   
4  CharlesCreativeContent      Shawn Charles               AMAZON  Boston, MA   

              email  hireable  \
0  brian@brianyu.me     False   
1               NaN     False   
2               NaN     False   
3               NaN      True   
4               NaN      True   

                                                 bio  public_repos  followers  \
0                    Software developer and educator            37      13500   
1  Creator of ESLint, independent software develo...            74      10529   
2     Smart Contract Engineer, Auditor, and Educator           279      10196   
3                        YouTube Content Creator :).            59       5944   
4        Software Engineer building Tech Communities            86       5063   

   following            created_at  
0         15  2015-11-29T07:25:29Z  
1          0  2008-12-05T17:18:56Z  
2         44  2019-08-19T14:13:41Z  
3          1  2013-12-25T19:49:26Z  
4       1098  2020-03-11T20:10:11Z

In [8]:
users['hireable'] = users['hireable'].fillna(False).astype(bool)

In [9]:
top5 = users.sort_values(by='followers', ascending=False).head()
print(', '.join(top5['login'].tolist()))

brianyu28, nzakas, PatrickAlphaC, KeithGalli, CharlesCreativeContent


#### Q2. Who are the 5 earliest registered GitHub users in Boston? List their login in ascending order of created_at, comma-separated.

In [10]:
users['created_at'] = pd.to_datetime(users['created_at'])

In [11]:
top_earliest = users.sort_values(by='created_at').head()
print(','.join(top_earliest['login'].tolist()))

evan,dpickett,tel,radical,joshuaclayton


#### Q3. What are the 3 most popular license among these users? Ignore missing licenses. List the license_name in order, comma-separated.

In [12]:
repos = pd.read_csv('repositories.csv')
repos.head()

login                  full_name            created_at  \
0  brianyu28        brianyu28/presenter  2024-11-24T03:57:34Z   
1  brianyu28         brianyu28/magnolia  2024-04-24T12:32:04Z   
2  brianyu28         brianyu28/jeopardy  2020-01-20T04:21:06Z   
3  brianyu28  brianyu28/classroometrics  2022-08-10T00:35:48Z   
4  brianyu28        brianyu28/brianyu28  2021-04-18T23:10:34Z   

   stargazers_count  watchers_count    language  has_projects  has_wiki  \
0                12              12  TypeScript          True     False   
1                 3               3      Python          True     False   
2                31              31  TypeScript          True      True   
3                18              18      Python          True      True   
4                12              12         NaN          True      True   

  license_name  
0     lgpl-2.1  
1     lgpl-2.1  
2          NaN  
3          NaN  
4          NaN

In [17]:
repos['license_name'].value_counts().head(3)

license_name
mit           10794
other          4868
apache-2.0     3787
Name: count, dtype: int64

#### Q4. Which company do the majority of these developers work at?

In [26]:
users['company'].value_counts().head(3)

company
NORTHEASTERN UNIVERSITY    19
GOOGLE                     12
MIT                        12
Name: count, dtype: int64

#### Q5. Which programming language is most popular among these users?

In [31]:
repos['language'].value_counts().head(3)

language
JavaScript    3642
Python        3006
Ruby           889
Name: count, dtype: int64

#### Q6. Which programming language is the second most popular among users who joined after 2020?


In [32]:
users_after_2020 = users[users['created_at'] > '2020-01-01']
users_after_2020.head()

login             name           company    location  \
4    CharlesCreativeContent    Shawn Charles            AMAZON  Boston, MA   
43              neuralmagic     Neural Magic               NaN      Boston   
88          Lanny-MacMillan  Lanny MacMillan             THRYV  Boston, Ma   
112        ibrahimgurhandev   Ibrahim Gurhan  RESILIENT CODERS   Boston MA   
130           jessicajaniuk   Jessica Janiuk            GOOGLE  Boston, MA   

    email  hireable                                                bio  \
4     NaN      True        Software Engineer building Tech Communities   
43    NaN     False  Neural Magic empowers developers to optimize a...   
88    NaN     False  Software Developer / Game Developer\r\nReact.j...   
112   NaN     False  Software Engineer at  Resilient Coders | Web D...   
130   NaN     False  This account is retired. Find Jessica at githu...   

     public_repos  followers  following                created_at  
4              86       5063       1098 2020-03-11 20:10:11+00:00  
43             62        855          0 2020-07-23 00:16:47+00:00  
88             25        493        830 2022-04-01 18:10:46+00:00  
112            36        378        798 2020-07-20 04:11:30+00:00  
130             6        341          0 2020-10-12 21:49:39+00:00

In [33]:
repos_2020 = repos[repos['login'].isin(users_after_2020['login'].tolist())]
repos_2020['language'].value_counts().head()

language
Python        58
JavaScript    32
TypeScript    24
HTML          13
Rust          11
Name: count, dtype: int64

#### Q7. Which language has the highest average number of stars per repository? 

In [42]:
# Calculate the average number of stars per language
avg_stars = repos.groupby('language')['stargazers_count'].mean()
# Find the language with the highest average stars
top_lang = avg_stars.idxmax()
top_stars = avg_stars.max()
print(top_lang, top_stars)

RMarkdown 787.0


#### Q8. Let's define leader_strength as followers / (1 + following). Who are the top 5 in terms of leader_strength? List their login in order, comma-separated.

In [43]:
users['leader_strength'] = users['followers'] / (1 + users['following'])
top5_lead = users.sort_values(by='leader_strength', ascending=False).head()
print(', '.join(top5_lead['login'].tolist()))

nzakas, nikomatsakis, ccoenraets, KeithGalli, rstudio


#### Q9. What is the correlation between the number of followers and the number of public repositories among users in Boston?

In [ ]:
# Compute correlation coefficient
correlation = users['followers'].corr(users['public_repos'])
correlation

0.15158303475607124

##### Solutoion 9:The correlation value is 0.1516, it means there is a weak positive correlation between the number of followers and the number of public repositories among users in Boston.

#### Q10. Does creating more repos help users get more followers? Using regression, estimate how many additional followers a user gets per additional public repository.

In [ ]:
import csv
followers = []
public_repos = []
with open('users.csv', 'r', encoding='utf-8') as file:
    reader = csv.DictReader(file)
    # Extract followers and public_repos Data
    for row in reader:
        followers_count = int(row['followers'])
        public_repos_count = int(row['public_repos'])
        followers.append(followers_count)
        public_repos.append(public_repos_count)
# Perform Linear Regression
if len(followers) > 1 and len(public_repos) > 1:
    slope, intercept = np.polyfit(public_repos, followers, 1)
    
    print(f"{slope:.3f}")
else:
    print("Error")

1.193


#### Q11. Do people typically enable projects and wikis together? What is the correlation between a repo having projects enabled and having wiki enabled?

In [49]:
if repos['has_projects'].dtype == 'object':
    repos['has_projects'] = repos['has_projects'].map({'true': True, 'false': False})
if repos['has_wiki'].dtype == 'object':
    repos['has_wiki'] = repos['has_wiki'].map({'true': True, 'false': False})
    
correlation = repos['has_projects'].corr(repos['has_wiki'])
    
print(round(correlation, 3))

0.275


#### Q12. Do hireable users follow more people than those who are not hireable?

In [50]:
hireable_avg_following = users[users['hireable'] == True]['following'].mean()
non_hireable_avg_following = users[users['hireable'] == False]['following'].mean()
difference = hireable_avg_following - non_hireable_avg_following
difference

112.88345285758281

#### Q13. Some developers write long bios. Does that help them get more followers? What's the correlation of the length of their bio (in Unicode characters) with followers? (Ignore people without bios)

In [53]:
import pandas as pd
from sklearn.linear_model import LinearRegression
# Load the CSV file into a DataFrame
users = pd.read_csv('users.csv')
# Filter users with non-empty 'bio' and create a new column 'bio_len' for the length of the bio
users_with_bio = users[(users['bio'].notna()) & (users['bio'] != '')].copy()
users_with_bio.loc[:, 'bio_len'] = users_with_bio['bio'].str.len()
# Prepare the data for linear regression
X = users_with_bio['bio_len'].values.reshape(-1, 1)
y = users_with_bio['followers']
# Perform linear regression
lr2 = LinearRegression()
lr2.fit(X, y)

# Print the coefficient
print(lr2.coef_[0])


-0.7698092960909222


#### Q14. Who created the most repositories on weekends (UTC)? List the top 5 users' login in order, comma-separated

In [55]:
import csv
from collections import Counter
from datetime import datetime

weekend_repo_counts = Counter()

with open('repositories.csv', 'r', encoding='utf-8') as file:
    reader = csv.DictReader(file)
    
    for row in reader:
        created_at = row.get('created_at', '')
        if created_at:
            created_date = datetime.fromisoformat(created_at[:-1])  
            
            if created_date.weekday() in [5, 6]:
                user_login = row['login']
                weekend_repo_counts[user_login] += 1  

top_users = weekend_repo_counts.most_common(5)

top_logins = [user[0] for user in top_users]

print(', '.join(top_logins))

cameronraysmith, berquist, burtbeckwith, jimkang, johnny-rice


#### Q15. Do people who are hireable share their email addresses more often?

In [56]:
fraction_hierable = users[users['hireable'] == True]['email'].notna().mean()
fraction_non_hierable = users[users['hireable'] == False]['email'].notna().mean()
diff = fraction_hierable - fraction_non_hierable
diff

0.08603546130494083

#### Q16. Let's assume that the last word in a user's name is their surname (ignore missing names, trim and split by whitespace.) What's the most common surname? (If there's a tie, list them all, comma-separated, alphabetically)

In [62]:
new_users = users[users['name'].notna()].copy()
# Extract the Surname (Last Word of Name)
new_users['surname'] = new_users['name'].str.split().str[-1].str.strip()
surname_counts = new_users['surname'].value_counts()

# Find the Most Common Surnames
max_count = surname_counts.max()
common_surnames = surname_counts[surname_counts == max_count].index.tolist()
common_surnames.sort()
print(','.join(common_surnames))

Williams,Zhang


In [ ]:
# Replaces missing values & converts everything to strings
users = users.fillna("").astype(str)
repos = repos.fillna("").astype(str)

# Normalizes boolean values to lowercase strings
repos['has_wiki'] = repos['has_wiki'].astype(str).replace({'True': 'true', 'False': 'false'})
repos['has_projects'] = repos['has_projects'].astype(str).replace({'True': 'true', 'False': 'false'})

# Saves the cleaned data for future use
users.to_csv('users1.csv', index=False)
repos.to_csv('repos1.csv', index=False)